storage/innobase/os/os0file.cc

```diff
@@ -4144,24 +4145,35 @@ bool fil_node_t::read_page0()
 
   if (!deferred)
   {
-    page_t *page= static_cast<byte*>(aligned_malloc(psize, psize));
-    if (os_file_read(IORequestRead, handle, page, 0, psize, nullptr)
-        != DB_SUCCESS)
+    page_t *apage= static_cast<byte*>(aligned_malloc(psize, psize));
+    if (os_file_read(IORequestRead, handle, apage, 0, psize, nullptr) !=
+        DB_SUCCESS)
     {
       sql_print_error("InnoDB: Unable to read first page of file %s", name);
-      aligned_free(page);
+   err_exit:
+      aligned_free(apage);
       return false;
     }
 
+    const page_t *page= apage;
+  retry:
     const ulint space_id= memcmp_aligned<2>
       (FIL_PAGE_SPACE_ID + page,
        FSP_HEADER_OFFSET + FSP_SPACE_ID + page, 4)
       ? ULINT_UNDEFINED
       : mach_read_from_4(FIL_PAGE_SPACE_ID + page);
     ulint flags= fsp_header_get_flags(page);
     const uint32_t size= fsp_header_get_field(page, FSP_SIZE);
+    if (!space_id && !flags && !size && dpage)
+    {
+    retry_dpage:
+      page= dpage;
+      dpage= nullptr;
+      goto retry;
+    }
```

Comment:
I get the idea that if we see a corrupt page then we continue with the double-write page if passed by the caller. Actually the function reads page-0 and fills up some attributes in fil_space_t. If the read page is corrupt the we use the double-write page passed by the caller as source.

Not something for the current patch but for future may be a better abstraction would be to separate out the reading and setting part. Something like ...

```
fil_space_t::set_page0_attributes(bool check_double_write) {

// 1. Read page0 from disk
page = os_file_read();

// 2. Attempt to initialize space from data page 
bool print_warning = !check_double_write;
bool success = init_from_page(page, print_warning);
if (success || !check_double_write) 
  return;

// 3. Attempt to initialize space from double write page
page = ... // Check and get from double-write pages
print_warning = true;
success = init_from_page(page, print_warning);
```

Reply:
That is definitely worth considering. At the same time we should try to merge `Datafile` to `fil_node_t` and `Tablespace` to `fil_space_t`; I think that they are duplicating some logic and causing some redundant system calls. One `open`, `fstat` and `pread` per file should really be enough.